src/libolm_migration.rs

```diff
@@ -0,0 +1,436 @@
+#[wasm_bindgen(getter_with_clone)]
+#[derive(Debug, Default)]
+pub struct BaseMigrationData {
+    /// The user id of the account owner.
+    #[wasm_bindgen(js_name = "userId")]
+    pub user_id: Option<UserId>,
+
+    /// The device ID of the account owner.
+    #[wasm_bindgen(js_name = "deviceId")]
+    pub device_id: Option<DeviceId>,
+
+    /// The pickled version of the Olm Account.
+    #[wasm_bindgen(js_name = "pickledAccount")]
+    pub pickled_account: String,
+
+    /// The backup version that is currently active.
+    #[wasm_bindgen(js_name = "backupVersion")]
+    pub backup_version: Option<String>,
+
+    /// The backup recovery key, as a base64-encoded string.
+    #[wasm_bindgen(js_name = "backupRecoveryKey")]
+    pub backup_recovery_key: Option<String>,
+
+    /// The private, base64-encoded, master cross-signing key.
+    #[wasm_bindgen(js_name = "crossSigningMasterKey")]
+    pub cross_signing_master_key: Option<String>,
+
+    /// The private, base64-encoded, self-signing key.
+    #[wasm_bindgen(js_name = "crossSigningSelfSigningKey")]
+    pub cross_signing_self_signing_key: Option<String>,
+
+    /// The private, base64-encoded, user-signing key.
+    #[wasm_bindgen(js_name = "crossSigningUserSigningKey")]
+    pub cross_signing_user_signing_key: Option<String>,
+}
+
+#[wasm_bindgen]
+impl BaseMigrationData {
+    /// Create a new `BaseMigrationData` with default values.
+    #[wasm_bindgen(constructor)]
+    pub fn new() -> Self {
+        Self::default()
+    }
+}
+
+#[wasm_bindgen]
+impl Migration {
+    /// Import the base dataset from a libolm-based setup to a vodozemac-based
+    /// setup stored in IndexedDB.
+    ///
+    /// Populates the user credentials, Olm account, backup data, etc. This is the
+    /// first step in the migration process. Once this base data is imported,
+    /// further data can be imported with {@link #migrateOlmSessions}, {@link
+    /// #migrateMegolmSessions}, and TODO room settings.
+    ///
+    /// # Arguments
+    ///
+    /// * `data` - The data to be migrated
+    /// * `pickle_key` - The libolm pickle key that was used to pickle the olm
+    ///   account objects.
+    /// * `store_handle` - A connection to the CryptoStore which will be used to
+    ///   store the vodozemac data.
+    #[wasm_bindgen(js_name="migrateBaseData")]
+    pub async fn migrate_base_data(
+        data: &BaseMigrationData,
+        pickle_key: Uint8Array,
+        store_handle: &StoreHandle,
+    ) -> Result<JsValue, JsError> {
+        migrate_base_data_to_store(data, &(pickle_key.to_vec()), store_handle.store.as_ref())
+            .await
+            .map_err(|e| JsError::from(&*e))?;
+        Ok(JsValue::UNDEFINED)
+    }
+}
+
+async fn migrate_base_data_to_store(
+    data: &BaseMigrationData,
+    pickle_key: &[u8],
+    store: &DynCryptoStore,
+) -> anyhow::Result<()> {
+    let user_id = data.user_id.clone().ok_or(anyhow!("User ID not specified"))?.inner;
```

Comment:
Nice tip about the `Option` extension, thanks!

> Alternatively, since the user_id is mandatory in the BaseMigrationData, let's make it a non Optional field?

Well... that would mean we'd have to pass in a UserID in the constructor of `BaseMigrationData`, and once we go down that path, pretty soon we get to a situation where the constructor takes 150 parameters. The whole point of `BaseMigrationData` is to avoid functions that take lots of parameters and instead allow the javascript side to use named properties.

Reply:
I think there's a middle ground where the two required parameters are provided (user id and device id), and then every other optional parameter can be set manually. But I won't insist since this is bindings code, as long as you're happy with the API shape here.

Reply:
Worth noting that `pickled_account` is also required: it's just that we don't have an explicit check for it here because it's not strongly typed (it defaults to an empty string, which is an invalid value), so strictly speaking that's three required parameters.

I think the current arrangement is the least bad solution in terms of the js-side API.